File: packages/runpane-py/src/runpane/releases.py
from __future__ import annotations

import json
import os
import re
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any, Dict

from .platforms import PanePlatform, arch_aliases, default_format, platform_param

GITHUB_API_BASE = "https://api.github.com/repos/greenfield-inc/Pane/releases"
DOWNLOAD_API_BASE = "https://runpane.com/api/download"


@dataclass
class ResolvedRelease:
    release: Dict[str, Any]
    artifact: Dict[str, Any]
    format: str
    preferred_download_url: str
    fallback_download_url: str
    checksum_url: str


def resolve_release(
    *,
    version: str,
    channel: str,
    source: str,
    platform: PanePlatform,
    format_name: str,
    target: str,
    fetch_timeout_seconds: float = 30,
) -> ResolvedRelease:
    release = fetch_release(version, timeout_seconds=fetch_timeout_seconds)
    selected_format = default_format(platform, target) if format_name == "auto" else format_name
    artifact = find_artifact(release, platform, selected_format)
    preferred = build_preferred_download_url(channel, source, platform, selected_format, release)
    tag_name = release["tag_name"]
    return ResolvedRelease(
        release=release,
        artifact=artifact,
        format=selected_format,
        preferred_download_url=preferred,
        fallback_download_url=artifact["browser_download_url"],
        checksum_url=f"https://github.com/greenfield-inc/Pane/releases/download/{tag_name}/SHA256SUMS.txt",
    )


def fetch_release(version: str, timeout_seconds: float = 30) -> Dict[str, Any]:
    normalized = "latest" if version == "latest" else f"tags/{version if version.startswith('v') else 'v' + version}"
    req = urllib.request.Request(
        f"{GITHUB_API_BASE}/{normalized}",
        headers={"Accept": "application/vnd.github+json", "User-Agent": "runpane-installer"},
    )
    with urllib.request.urlopen(req, timeout=timeout_seconds) as response:
        release = json.loads(response.read().decode("utf-8"))

    if release.get("draft") or release.get("prerelease"):
        raise RuntimeError(f"Release {release.get('tag_name')} is not a stable public release.")
    return release


def find_artifact(release: Dict[str, Any], platform: PanePlatform, format_name: str) -> Dict[str, Any]:
    assets = release.get("assets") or []
    candidates = [
        asset for asset in assets
        if matches_format(asset["name"], format_name) and matches_platform(asset["name"], platform)
    ]
    aliases = arch_aliases(platform)
    for asset in candidates:
        lower = asset["name"].lower()
        if any(alias.lower() in lower for alias in aliases):
            return asset
    for asset in candidates:
        if "universal" in asset["name"].lower():
            return asset
    if candidates:
        return candidates[0]
    names = ", ".join(asset["name"] for asset in assets) or "no assets"
    raise RuntimeError(f"No Pane {format_name} asset found for {platform.os}/{platform.arch}. Assets: {names}")


def artifact_file_name(url_or_name: str) -> str:
    return os.path.basename(url_or_name.split("?", 1)[0])


def build_preferred_download_url(
    channel: str,
    source: str,
    platform: PanePlatform,
    format_name: str,
    release: Dict[str, Any],
) -> str:
    query = urllib.parse.urlencode({
        "platform": platform_param(platform),
        "arch": platform.arch,
        "format": format_name,
        "version": release["tag_name"],
        "channel": channel,
        "source": source,
    })
    return f"{DOWNLOAD_API_BASE}?{query}"


def matches_format(name: str, format_name: str) -> bool:
    lower = name.lower()
    if format_name == "appimage":
        return lower.endswith(".appimage")
    return lower.endswith(f".{format_name}")


def matches_platform(name: str, platform: PanePlatform) -> bool:
    lower = name.lower()
    if platform.os == "darwin":
        return "macos" in lower or "darwin" in lower or "mac" in lower
    if platform.os == "win32":
        return "windows" in lower or re.search(r"(?:^|[._-])win(?:32|64)?(?:[._-]|$)", lower) is not None
    return "linux" in lower

File: packages/runpane/src/releases.ts
import path from 'path';
import { boundary, decodeBoundary } from './boundaryDecoder';
import type { ArtifactFormat } from './commands';
import type { BoundarySchema } from './boundaryDecoder';
import { archAliases, defaultFormat, platformParam, type PanePlatform } from './platform';

const GITHUB_API_BASE = 'https://api.github.com/repos/greenfield-inc/Pane/releases';
const DOWNLOAD_API_BASE = 'https://runpane.com/api/download';

interface GitHubReleaseAsset {
  name: string;
  browser_download_url: string;
}

interface GitHubRelease {
  tag_name: string;
  name: string;
  body: string;
  html_url: string;
  published_at: string;
  prerelease: boolean;
  draft: boolean;
  assets?: GitHubReleaseAsset[];
}

const githubReleaseSchema: BoundarySchema<GitHubRelease> = boundary.object({
  tag_name: boundary.string,
  name: boundary.string,
  body: boundary.string,
  html_url: boundary.string,
  published_at: boundary.string,
  prerelease: boundary.boolean,
  draft: boundary.boolean,
  assets: boundary.optional(boundary.array(boundary.object({
    name: boundary.string,
    browser_download_url: boundary.string,
  }))),
});

export interface ResolvedRelease {
  release: GitHubRelease;
  artifact: GitHubReleaseAsset;
  format: Exclude<ArtifactFormat, 'auto'>;
  preferredDownloadUrl: string;
  fallbackDownloadUrl: string;
  checksumUrl: string;
}

export interface ResolveReleaseOptions {
  version: string;
  channel: 'stable' | 'nightly';
  source: 'npm' | 'pip';
  platform: PanePlatform;
  format: ArtifactFormat;
  target: 'client' | 'daemon';
  fetchTimeoutMs?: number;
}

export async function resolveRelease(options: ResolveReleaseOptions): Promise<ResolvedRelease> {
  const release = await fetchRelease(options.version, options.fetchTimeoutMs);
  const format = options.format === 'auto' ? defaultFormat(options.platform, options.target) : options.format;
  const artifact = findArtifact(release, options.platform, format);
  const preferredDownloadUrl = buildPreferredDownloadUrl(options, format, release);

  return {
    release,
    artifact,
    format,
    preferredDownloadUrl,
    fallbackDownloadUrl: artifact.browser_download_url,
    checksumUrl: `https://github.com/greenfield-inc/Pane/releases/download/${release.tag_name}/SHA256SUMS.txt`
  };
}

export async function fetchRelease(version: string, timeoutMs?: number): Promise<GitHubRelease> {
  const normalized = version === 'latest' ? 'latest' : version.startsWith('v') ? `tags/${version}` : `tags/v${version}`;
  const controller = timeoutMs ? new AbortController() : undefined;
  const timeout = controller
    ? setTimeout(() => controller.abort(), timeoutMs)
    : undefined;

  try {
    const response = await fetch(`${GITHUB_API_BASE}/${normalized}`, {
      headers: {
        accept: 'application/vnd.github+json',
        'user-agent': 'runpane-installer'
      },
      signal: controller?.signal
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch Pane release ${version}: ${response.status} ${response.statusText}`);
    }

    const release = decodeBoundary(await response.json(), githubReleaseSchema);
    if (release.draft || release.prerelease) {
      throw new Error(`Release ${release.tag_name} is not a stable public release.`);
    }
    return release;
  } catch (error) {
    if (error instanceof Error && error.name === 'AbortError') {
      throw new Error(`Timed out fetching Pane release ${version} after ${timeoutMs}ms.`);
    }
    throw error;
  } finally {
    if (timeout) {
      clearTimeout(timeout);
    }
  }
}

export function findArtifact(
  release: GitHubRelease,
  platform: PanePlatform,
  format: Exclude<ArtifactFormat, 'auto'>
): GitHubReleaseAsset {
  const assets = release.assets ?? [];
  const aliases = archAliases(platform);
  const candidates = assets.filter((asset) => matchesFormat(asset.name, format) && matchesPlatform(asset.name, platform));
  const exact = candidates.find((asset) => aliases.some((alias) => lowerName(asset).includes(alias.toLowerCase())));
  const universal = candidates.find((asset) => lowerName(asset).includes('universal'));
  const selected = exact ?? universal ?? candidates[0];

  if (!selected) {
    const assetNames = assets.map((asset) => asset.name).join(', ') || 'no assets';
    throw new Error(`No Pane ${format} asset found for ${platform.os}/${platform.arch} in ${release.tag_name}. Assets: ${assetNames}`);
  }

  return selected;
}

export function artifactFileName(urlOrName: string): string {
  return path.basename(urlOrName.split('?')[0]);
}

function buildPreferredDownloadUrl(
  options: ResolveReleaseOptions,
  format: Exclude<ArtifactFormat, 'auto'>,
  release: GitHubRelease
): string {
  const params = new URLSearchParams({
    platform: platformParam(options.platform),
    arch: options.platform.arch,
    format,
    version: release.tag_name,
    channel: options.channel,
    source: options.source
  });
  return `${DOWNLOAD_API_BASE}?${params.toString()}`;
}

function matchesFormat(name: string, format: Exclude<ArtifactFormat, 'auto'>): boolean {
  const lower = name.toLowerCase();
  if (format === 'appimage') return lower.endsWith('.appimage');
  return lower.endsWith(`.${format}`);
}

function matchesPlatform(name: string, platform: PanePlatform): boolean {
  const lower = name.toLowerCase();
  if (platform.os === 'darwin') return lower.includes('macos') || lower.includes('darwin') || lower.includes('mac');
  if (platform.os === 'win32') return lower.includes('windows') || /(?:^|[._-])win(?:32|64)?(?:[._-]|$)/.test(lower);
  return lower.includes('linux');
}

function lowerName(asset: GitHubReleaseAsset): string {
  return asset.name.toLowerCase();
}
